pybamm/experiment/steps.py

```diff
@@ -0,0 +1,214 @@
+#
+# Public functions to create steps for use in an experiment.
+#
+from ._steps_util import _Step, _convert_electric, _examples
+
+
+def string(string, **kwargs):
+    """
+    Create a step from a string.
+
+    Parameters
+    ----------
+    string : str
+        The string to parse. Each operating condition should
+        be of the form "Do this for this long" or "Do this until this happens". For
+        example, "Charge at 1 C for 1 hour", or "Charge at 1 C until 4.2 V", or "Charge
+        at 1 C for 1 hour or until 4.2 V at 25oC". The instructions can be of the form
+        "(Dis)charge at x A/C/W", "Rest", or "Hold at x V until y A". The running
+        time should be a time in seconds, minutes or hours, e.g. "10 seconds",
+        "3 minutes" or "1 hour". The stopping conditions should be
+        a circuit state, e.g. "1 A", "C/50" or "3 V".
+    **kwargs
+        Any other keyword arguments are passed to the :class:`pybamm.experiment.Step`
+        class.
+
+    Returns
+    -------
+    :class:`pybamm.experiment.Step`
+        A step parsed from the string.
+    """
+    if not isinstance(string, str):
+        raise TypeError("Input to experiment.string() must be a string")
+
+    if "period)" in string:
+        raise ValueError(
+            "Period must be specified as a keyword argument instead of in the string"
+        )
+
+    if "oC" in string:
+        raise ValueError(
+            "Temperature must be specified as a keyword argument "
+            "instead of in the string"
+        )
+
+    # Save the original string
+    description = string
+
+    # extract termination condition based on "until" keyword
+    if "until" in string:
+        # e.g. "Charge at 4 A until 3.8 V"
+        string, termination = string.split(" until ")
+        # sometimes we use "or until" instead of "until", so remove "or"
+        string = string.replace(" or", "")
+    else:
+        termination = None
+
+    # extract duration based on "for" keyword
+    if "for" in string:
+        # e.g. "Charge at 4 A for 3 hours"
+        string, duration = string.split(" for ")
+    else:
+        duration = None
+
+    if termination is None and duration is None:
+        raise ValueError(
+            "Operating conditions must contain keyword 'for' or 'until'. "
+            f"For example: {_examples}"
+        )
+
+    # read remaining instruction
+    if string.startswith("Rest"):
+        typ = "current"
+        value = 0
+    else:
+        # split by what is before and after "at"
+        # e.g. "Charge at 4 A" -> ["Charge", "4 A"]
+        # e.g. "Discharge at C/2" -> ["Discharge", "C/2"]
+        instruction, value_string = string.split(" at ")
+        if instruction == "Charge":
+            sign = -1
+        elif instruction in ["Discharge", "Hold"]:
+            sign = 1
+        else:
+            raise ValueError(
+                "Instruction must be 'discharge', 'charge', 'rest', or 'hold'. "
```

Comment:
Do we want a separate error message for `"run"`, as it has now been deprecated? I mean, it will be caught by this error, but it is less informative.